pkg/operator/status.go

```diff
@@ -98,55 +94,87 @@ func isDeploymentStatusComplete(deploy *appsv1.Deployment) bool {
-			c.setProgressingStatus(operatorConfigCopy, statusMsg)
+			statusMsg = fmt.Sprintf("Deployment %s does not have available replicas", dep)
+			c.setProgressingTrue(operatorConfigCopy, reason, statusMsg)
+			// If there isn't at least one replica from each deployment, Available=False
+			c.setAvailableFalse(operatorConfigCopy, reason, statusMsg)
 			return false, nil
 		}
 		existingDeploymentsAndReplicas++
 
-		if !isDeploymentStatusComplete(existing) {
-			glog.Infof("The deployment %s has not completed", dep)
-			statusMsg := fmt.Sprintf("Deployment %s has not completed", dep)
-			c.setProgressingStatus(operatorConfigCopy, statusMsg)
-			return false, nil
-		}
 		if isDeploymentStatusAvailableAndUpdated(existing) {
 			glog.Infof("Deployment %s is available and updated", dep)
-			ready++
+			version_ready++
+			if isDeploymentStatusComplete(existing) {
+				glog.Infof("Deployment %s has desired replicas.", dep)
+				deployment_complete++
+			} else {
+				statusMsg = fmt.Sprintf("Deployment %s is creating replicas.", dep)
+			}
+		} else {
+			statusMsg = fmt.Sprintf("Deployment %s is updating", dep)
 		}
 	}
-	// set Available if replica instances are created.
-	// report version if all deployments are available and updated
-	if ready == len(deployments) {
-		glog.Infof("All deployments managed by service-ca-operator are available and updated")
-		c.setAvailableStatus(operatorConfigCopy)
+	// Available, Updated, and Ready to report version:
+	// Here, ready to report version and set Available=True and
+	// set Progressing=False because all deployments, replicas exist
+	// and all instances are updated, no previous deployment instances exist
+	if deployment_complete == len(deployments) {
+		reason := "ManagedDeploymentsCompleteAndUpdated"
+		c.setAvailableTrue(operatorConfigCopy, reason)
+		c.setProgressingFalse(operatorConfigCopy, reason, "All service-ca-operator deployments updated")
+		return true, nil
+	}
+	// Step down in readiness:
+	// Here, ready to report a version and set Available=True
+	// because all deployments exist and at least 1 replica exists per deployment,
+	// and there are no instances from previous deployments.
+	// Progressing will be true because deployments aren't complete.
+	if version_ready == len(deployments) {
+		reason := "ManagedDeploymentsAvailableAndUpdated"
+		c.setAvailableTrue(operatorConfigCopy, reason)
+		c.setProgressingTrue(operatorConfigCopy, reason, statusMsg)
 		return true, nil
 	}
-	// report Available is deployments and replicas exist, but don't report version
-	// if there are replica instances remaining from previous deployment
+	// Further step down in readiness:
+	// Here, ready to report Available=True
+	// because all deployments exist and at least 1 replicas exists per deployment.
+	// Don't report version here because
+	// there are replica instances remaining from previous deployment
+	// Progressing will be true here
```

Comment:
`all deployments have at least 1 replica, but some are of previous versions - don't report new version`?

Reply:
:+1: